Tests: Update, use makem.sh, GitHub Actions

alphapapa opened the pull request:
* (.github/workflows/test.yml, Makefile, makem.sh): This adds GitHub
  Actions support using makem.sh.
* test/quelpa-test.el: Fix requirements, and don't bootstrap Quelpa by
  downloading it from GitHub's master raw link over HTTPS, which
  defeats the purpose of testing the local HEAD and makes it not
  clean/safe to test locally.

This does not intend to alter the existing CI infrastructure.  However, I don't know if removing the "bootstrap" part will break the Travis CI.  If it does, it should be easy to fix.

The main benefit of this PR is making it safer, easier, and faster to run linting and tests locally.

You can see an example of the GitHub Actions run here: https://github.com/alphapapa/quelpa/runs/375453793  It shows that the tests pass until 26.1.

alphapapa commented:
> I checked the run and couldn't find the system tests (emacs running the `.emacs` file). Did I miss something?

No, `makem.sh` runs Buttercup and ERT tests automatically.  Those tests in `ci/.emacs` should probably be rewritten as ERT tests.

steckerhalter commented:
> > I checked the run and couldn't find the system tests (emacs running the `.emacs` file). Did I miss something?
> 
> No, `makem.sh` runs Buttercup and ERT tests automatically. Those tests in `ci/.emacs` should probably be rewritten as ERT tests.

hmm, ok. well, the purpose of running a full config with `.emacs` is also to test the whole bootstrapping process too. I'm not sure if this can be done with ERT tests. what do you think?

at any rate it was quite a good way to know if something was broken or not.

alphapapa commented:
I'm not very familiar with the bootstrapping part.  If it's still necessary, and if it can't be written as an ERT test, then we could run that test outside of `makem.sh` by adding it to the `test.yml` file and/or the `makem.sh/Makefile` file (which I didn't include in this PR, since it would conflict with the existing Makefile).

That could also be done separately from this PR, although the code that was removed from `quelpa-test.el` will need to be reimplemented separately, so that it won't run when the ERT test file is loaded.

In any case, the `Makefile` and `test.yml` from `makem.sh`'s repo are intended to be templates that devs can modify to fit their packages as needed.  `makem.sh` is the part that isn't intended to be modified (of course, it could be, but I'd prefer that issues with it be reported upstream so I can enhance it by default).

steckerhalter commented:
if it can be done as you say and the existing `.emacs` can be run somehow with `exec` or so then I'd say it is well and we could drop Travis... for this PR maybe it is enough to fix the error with emacs 26 and the rest can be done in another PR?

kiennq commented on bootstrap.el:
```diff
@@ -1,32 +0,0 @@
-
-;;; Code:
-
-(require 'package)
```
Deleting this file is breaking change. Lot of user who doesn't use Melpa will have to update their script to not use `bootstrap.el`. Perhaps you may want to update README or not delete this file

alphapapa replied:
Thanks, I misunderstood and thought it was only used for testing.

I would prefer to document that users should "bootstrap" manually by copying `quelpa.el` into their Emacs config rather than running `bootstrap.el`.  Is that okay with you, or do you prefer to keep it?

kiennq replied:
Copying `quelpa.el` works for me too. In that case, please also document it as it's breaking change. I would imagine that a lot of people that's using `bootstrap.el` suddenly got it stop working.
Nit: here is how I configure my quelpa from custom source. Perhaps we can adapt it and replace the part of using `bootstrap.el` in README file with that?
``` elisp
(unless (package-installed-p 'quelpa)
    (with-temp-buffer
      (url-insert-file-contents "https://raw.githubusercontent.com/quelpa/quelpa/master/quelpa.el")
      (eval-buffer)
      (quelpa-self-upgrade)))
```

alphapapa commented:
@steckerhalter I rewrote the `.emacs` tests into the `quelpa-tests.el` file.  You can see the last test runs here: https://github.com/alphapapa/quelpa/actions/runs/31478221  Most of the tests pass, but a few don't.  

* The `file` test was apparently designed to run on your local system where you have Helm checked out in a directory; we could fix that without too much trouble.
* The `url` test fails mysteriously.  If I run the code manually, locally, it seems to work, but the test doesn't pass when run from ERT.
* The `bzr` test will need `bzr` to be installed, which should be fixable in the `test.yml` file.  I'll take care of that later.
* The `stable` test fails because of a missing function.  I'm not sure what's going on there.  The test probably needs to be improved to compare the package versions before and after upgrading, anyway.

I think this is close to being ready to merge, if you approve of the general direction of the changes. 
 Please let me know what you think.

steckerhalter commented:
I'd suggest to keep `bootstrap.el` for now (but fix the quelpa url in there :) as not to break existing configs but add the relevant code to the README so `bootstrap.el` will not be needed anymore in the future (NEWS entry probably)

I'm not sure anymore if Windows is tested even. I can't find anything about it just now. I don't think it's very important so we can drop that if it is an issue.

You suggested that running the init could be done with an `exec` option, right? Maybe this would be easier for now and porting the tests ERT could be done in another PR?

steckerhalter commented:
Thinking about it, I'd say to have a system test with emacs bootstrapping quelpa (even without `bootstrap.el`) and running some quelpa commands would be still good to have (system tests can trigger different errors sometimes). Being able to do this with `exec` would be nice.

alphapapa commented:
> I'd suggest to keep `bootstrap.el` for now (but fix the quelpa url in there :) as not to break existing configs but add the relevant code to the README so `bootstrap.el` will not be needed anymore in the future (NEWS entry probably)

Ok, I'll look at that some more.

> I'm not sure anymore if Windows is tested even. I can't find anything about it just now. I don't think it's very important so we can drop that if it is an issue.

Sounds good, that will simplify it a lot.

> You suggested that running the init could be done with an `exec` option, right? Maybe this would be easier for now and porting the tests ERT could be done in another PR?

That shouldn't be necessary since the tests are now in ERT and `makem.sh` loads `quelpa.el` from the local repository into a sandbox (i.e. an empty Emacs config).  I think that downloading `quelpa.el` in the test suite doesn't make sense anyway, because that file is different from the one in the local system, i.e. the one that needs to be tested.

Since the tests have been moved to ERT, there shouldn't be a need for that `exec` trick I mentioned.

steckerhalter commented:
It is using the local `quelpa.el` in the test suite now, not a remote `quelpa.el`. If those tests run in ERT it's ok too. If we run emacs with a real world use case how quelpa is invoked and used in a normal init file it can help us to discover issues that are not visible in unit tests, but it's not a necessity so we can do it either way or both ways.

alphapapa commented:
@steckerhalter I converted the readme to an Org file and rewrote some of the installation and usage instructions.  Please let me know if it meets with your approval.  If so, I think this branch can be merged.  Thanks.

steckerhalter commented:
> @steckerhalter I converted the readme to an Org file and rewrote some of the installation and usage instructions. Please let me know if it meets with your approval. If so, I think this branch can be merged. Thanks.

Nice, thanks. It looks good to me.

steckerhalter commented:
It looks like all the tests fail on Github Actions here. Do you want to merge this with the failing tests?

alphapapa commented:
I see 4 out of 15 ERT tests failing.  One of them should be fixable by installing bzr in the CI setup.  The others I'm not sure about; one of them, at least, passes when run interactively but fails in batch mode, which I can't explain.  One of them may be exposing a bug in Quelpa.  They will require further investigation, but I think they can be fixed later rather than before this is merged.

alphapapa commented:
This one seems to be exposing a bug in Quelpa, I think: https://github.com/alphapapa/quelpa/runs/414613881#step:7:530

This one shows failure here https://github.com/alphapapa/quelpa/runs/414613881#step:7:650 even though the earlier output shows that the package was installed: https://github.com/alphapapa/quelpa/runs/414613881#step:7:629  I can't explain that.

This one fails because it was designed to run on (I guess) your personal system: https://github.com/alphapapa/quelpa/runs/414613881#step:7:104  It should be easy enough to fix in the CI by manually checking out the Helm repo and adjusting the path in the ERT test.  However, that means the ERT test will not pass on local developer systems.  Maybe that's acceptable.  Maybe we could disable the test except when it's running on CI.

I fixed the bzr test.

Let me know what you think.  Thanks.

steckerhalter commented:
Ok, can you disable the tests that fail for now? I'd prefer to have a build that gives at least some clear indication if everything is on track or not.

steckerhalter commented:
> This one seems to be exposing a bug in Quelpa, I think: https://github.com/alphapapa/quelpa/runs/414613881#step:7:530

It looks like the checkout method (git, svn ...) was not set here. The function should be e.g. `quelpa-build--checkout-git` (this is the from the original melpa code)

alphapapa commented:
Ok, the most recent commit makes tests pass (including those expected failures) back to Emacs 25.1.  

24.5 and earlier fail because something (possibly just a linter, but maybe also one of the packages installed while testing) requires 25.1 or later.  24.5 was released in April of 2015, so I'd be in favor of not bothering to test Emacs 24.x releases on CI.

steckerhalter commented:
yeah, I was thinking maybe the support for emacs 24 could be removed altogether

alphapapa commented:
Agreed!

steckerhalter commented:
@tsuu32 sorry, I think the idea was to [replace the bootstrapping snippet with something like that](https://github.com/quelpa/quelpa/pull/159#discussion_r370985084) (as suggested by @kiennq):

```elisp
(unless (package-installed-p 'quelpa)
    (with-temp-buffer
      (url-insert-file-contents "https://raw.githubusercontent.com/quelpa/quelpa/master/quelpa.el")
      (eval-buffer)
      (quelpa-self-upgrade)))
```
I'm going to document this now in the README and will re-add a `boostrap.el` that will only print a message to inform people to migrate away from the separate file.

tsuu32 commented:
@steckerhalter thanks for updating README. OK, I'll use it.
And @alphapapa @kiennq thanks for maintaining and updating quelpa!
